src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,683 @@
+                is_forbidden = true;
+                break;
+            }
+        }
+        if (is_forbidden) {
+            continue;
+        }
+
+        std::vector<Literal> conclusion = {l};
+        std::vector<Literal> premises = {};
+        Gfd gfd = Gfd(pattern, premises, conclusion);
+        if (Validate(graph, gfd, embeddings, sigma)) {
+            rules.push_back({{}, l});
+            continue;
+        }
+        std::vector<Literal> notsatisfied = current;
+        HGenerate(notsatisfied, l, rules, graph, pattern, embeddings, forbidden_rules, sigma);
+    }
+    return rules;
+}
+
+void AddCompacted(std::vector<SingleRule>& rules, std::vector<Gfd>& result, graph_t& pattern,
+                  std::vector<Rules>::iterator& forbidden_rules) {
+    if (!rules.empty()) {
+        std::sort(rules.begin(), rules.end());
+        std::vector<Literal> prev = rules.front().first;
+        std::vector<Literal> conclusion = {};
+        for (auto& [premises, l] : rules) {
+            if (prev != premises) {
+                Gfd gfd = Gfd(pattern, prev, conclusion);
+                result.push_back(gfd);
+                forbidden_rules->push_back({prev, conclusion});
+                prev = premises;
+                conclusion = {l};
+            } else {
+                conclusion.push_back(l);
+            }
+        }
+        Gfd gfd = Gfd(pattern, prev, conclusion);
+        result.push_back(gfd);
+        forbidden_rules->push_back({prev, conclusion});
+    }
+}
+
+void HorizontalSpawn(graph_t const& graph, std::size_t const& sigma,
+                     std::vector<graph_t> const& patterns,
+                     std::vector<Embeddings> const& embeddings_set,
+                     std::vector<Rules>& forbidden_rules_set,
+                     std::map<std::string, Info> const& attrs_info, std::vector<Gfd>& result) {
+    for (std::size_t i = 0; i < patterns.size(); ++i) {
+        auto pattern = patterns.at(i);
+        auto embeddings = embeddings_set.at(i);
+        auto forbidden_rules = std::next(forbidden_rules_set.begin(), i);
+
+        std::vector<Literal> literals = GenerateLiterals(pattern, attrs_info);
+        std::vector<SingleRule> rules =
+                GenerateRules(literals, graph, pattern, embeddings, forbidden_rules, sigma);
+        AddCompacted(rules, result, pattern, forbidden_rules);
+    }
+}
+
+class CompareCallback {
+private:
+    graph_t const& query_;
+    Embedding& iso_;
+    bool& res_;
+
+public:
+    CompareCallback(graph_t const& query_, Embedding& iso_, bool& res_)
+        : query_(query_), iso_(iso_), res_(res_) {}
+
+    template <typename CorrespondenceMap1To2, typename CorrespondenceMap2To1>
+    bool operator()(CorrespondenceMap1To2 f, CorrespondenceMap2To1) const {
+        BGL_FORALL_VERTICES_T(u, query_, graph_t) {
+            iso_.emplace(u, get(f, u));
+        }
+        res_ = true;
+        return false;
```

Comment:
The callback is needed to search for isomorphisms. The `operator()` function is called if an embedding was found, and also returns true if the search for isomorphisms should be continued, and false otherwise. In this case, it is enough to find at least one isomorphism, so I return false on the first call
To understand whether an embedding was found or not, I use the `res_` variable